dataset/include/scipp/dataset/dataset_util.h

```diff
@@ -8,15 +8,67 @@
 
 namespace scipp::dataset {
 
-/// Iteration helper for generic code supporting data array and dataset.
-///
-/// Example:
-///     for (const auto &[name, data] : iter(dataarray_or_dataset)) {}
-template <class T> static decltype(auto) iter(T &d) {
-  if constexpr (std::is_same_v<T, Dataset>)
-    return d;
-  else
-    return d.iterable_view();
+[[noreturn]] inline void throw_nesting_error(const DataArray &) {
+  throw std::invalid_argument("Cannot assign DataArray, the right hand side "
+                              "contains a reference to the left hand side. "
+                              "Reference cycles are not allowed.");
+}
+
+[[noreturn]] inline void throw_nesting_error(const Dataset &) {
+  throw std::invalid_argument("Cannot assign Dataset, the right hand side "
+                              "contains a reference to the left hand side. "
+                              "Reference cycles are not allowed.");
+}
+
+template <class L, class R>
+void check_nested_in_assign(const L &lhs, const R &rhs);
+
+template <class T>
+void check_nested_in_assign(const T &lhs, const Variable &rhs) {
+  if (rhs.dtype() == dtype<DataArray>) {
+    for (const auto &nested : rhs.values<DataArray>()) {
+      check_nested_in_assign(lhs, nested);
+    }
+  } else if (rhs.dtype() == dtype<Dataset>) {
+    for (const auto &nested : rhs.values<Dataset>()) {
+      check_nested_in_assign(lhs, nested);
+    }
+  }
```

Comment:
Do we also need to check the buffer of a binned variable here?

Reply:
Correct me if I'm wrong but Variables of Variables cannot be binned. So the only case where we would need to check the buffer is if that is a DataArray. The only way to get a cycle in this case would be trough
```.py
da = get_binned_data_array()
var = sc.scalar(da)
da.attrs['attr'] = var
```
which requires a check in `Dict` and not `Variable`.

Reply:
Not possible: 
```.py
---------------------------------------------------------------------------
DTypeError                                Traceback (most recent call last)
<ipython-input-14-3a82f34c1694> in <module>
----> 1 da.bins.coords['abc'] = var

DTypeError: Expected item dtype VariableView, got DataArray.
```